### PR TITLE
Allow filtering shipping methods by product groups

### LINKED ISSUE #837: Shipment method based on product groups?
I have this request from a customer of mine:

1. if the order total is below X: use shipping method M1 (which costs P1)
2. if the order total is X or more: use shipping method M2 (which costs P2)

BUT

3. if the order contains products belonging to certain groups, we should use shipping method M3 (which costs P3) - regardless of the order amount.

Community Store already allows implementing 1. and 2.

But I don't see a way to implement 3.
That's correct? If so, I'd submit a pull request that adds these two info to the shipping methods:
a. apply this shipping method only for orders that don't contain any product belonging to one or more of the following groups: ...
b. apply this shipping method only for orders that contain one or more product belonging to one or more of the following groups: ...

> [!NOTE]
> If you want to know the reason of this request:
> - for orders up to 99.99 €, the shipment costs 7 €
> - for orders from 100 €, the shipment is free
> - orders containing products discounted by 50% or 70%, the shipment costs 7€ for any order amount

### PR DESCRIPTION
**Pull request by mlocati**
Close #837

### CODE REVIEW

File: src/CommunityStore/Shipping/Method/ShippingMethod.php [hunk 1]
```diff
@@ -315,10 +338,26 @@ public static function getEligibleMethods()
...
             if ($method->getShippingMethodTypeMethod()->isEligible()) {
+                $criteria = $method->getProductGroupsCriteria();
+                if ($criteria) {
+                    if ($criteriaService === null) {
+                        $criteriaService = app(Criteria::class);
+                    }
+                    if ($products === null) {
+                        $em = app(EntityManagerInterface::class);
+                        $products = [];
+                        foreach (Cart::getCart() as $entry) {
```
Comment (mlocati) [inline src/CommunityStore/Shipping/Method/ShippingMethod.php:349]: This is the right way to get the items in the current cart, isn't it?

Comment (Mesuva) [reply src/CommunityStore/Shipping/Method/ShippingMethod.php:349]: Yes. It returns an array of items (so perhaps not the cleanest), but that would be the way to do it.

File: single_pages/dashboard/store/settings/shipping.php [hunk 2]
```diff
@@ -70,20 +78,69 @@
...
                                 </select>
                             </div>
                         </div>
                     </div>
                 </div>
+
+                <div class="row">
+                    <div class="col-md-12">
+                        <div class="form-group">
+                            <?= $form->label('methodProductGroupsCriteria', t('Applicability based on product groups')) ?>
```
Comment (mlocati) [inline single_pages/dashboard/store/settings/shipping.php:92]: @Mesuva can you think about a better wording?

Comment (mlocati) [reply single_pages/dashboard/store/settings/shipping.php]: I like your alternative!

**Pull request merged**